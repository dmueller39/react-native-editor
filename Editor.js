// @flow
import React, { Component } from 'react';

import { StyleSheet, KeyboardAvoidingView } from 'react-native';

import _ from 'underscore';
import {
  type StyleObj,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes';

import Buffer from './Buffer';
import ControlBar from './ControlBar';
import TextInputBar from './TextInputBar';
import text, { replaceRange } from './text';

import { COMMANDS } from './constants';

import { type RawLocation } from './types';

import {
  getNextLocation,
  getPreviousLocation,
  getDataWithReplaceWord,
} from './util';

const styles = StyleSheet.create({
  buffer: {
    overflow: 'hidden',
  },
  controlBar: {
    height: 44,
  },
});

const defaultProps = {
  data: '',
  onUpdateData: () => {},
};

type Props = {
  dimensions: {
    width: number,
  },
  data: string,
  onUpdateData: () => void,
  style?: ?StyleObj,
};

type State = {
  data?: string,
  selectedWord?: ?string,
  selectedLocation?: ?RawLocation,
  selectedLineIndex?: ?number,
  command?: ?string,
  replacementWord?: ?string,
  goToLineText?: ?string,
  selectedLineTimestamp?: ?number,
};

type DefaultProps = typeof defaultProps;

export default class Editor extends Component<DefaultProps, Props, State> {
  static defaultProps = defaultProps;

  constructor({ data }: Props) {
    super();
    this.state = { data };
  }

  state: State = {};

  bufferRef: ?Buffer = null;

  captureBufferRef = (buffer: Buffer) => {
    this.bufferRef = buffer;
  };

  componentDidUpdate() {
    this.props.onUpdateData(this.state.data);
  }

  onChangeData = () => {};

  onPressNextSelection() {
    const { data, selectedWord } = this.state;
    let { selectedLocation } = this.state;
    if (selectedWord == null || selectedLocation == null || data == null) {
      return;
    }
    selectedLocation = getNextLocation(selectedWord, selectedLocation, data);
    this.setState({ selectedLocation });
  }

  onPressPreviousSelection() {
    const { data, selectedWord } = this.state;
    let { selectedLocation } = this.state;
    if (selectedWord == null || selectedLocation == null || data == null) {
      return;
    }
    selectedLocation = getPreviousLocation(
      selectedWord,
      selectedLocation,
      data
    );
    this.setState({ selectedLocation });
  }

  onPressNextLine() {
    const { data } = this.state;
    let { selectedLineIndex } = this.state;
    if (selectedLineIndex == null || data == null) {
      return;
    }
    selectedLineIndex = Math.min(
      selectedLineIndex + 1,
      data.split('\n').length - 1
    );
    this.setState({ selectedLineIndex });
  }

  onPressInsertDone() {
    const buffer = this.bufferRef;
    if (buffer == null) {
      return;
    }

    const initialData = this.state.data;

    if (initialData == null) {
      return;
    }

    const edits = buffer.getEdits();

    // TODO clean this section up, and figure out exactly what is going on with
    // ./text
    const textState = edits.reduce(
      (textState: { data: string }, edit) => {
        const action = replaceRange(
          textState.data,
          edit.start,
          edit.end,
          edit.replacement
        );
        return text(textState, action);
      },
      { data: initialData }
    );

    const selectedLineIndex = buffer.getEditingLineIndex();

    this.setState({
      data: textState.data,
      command: COMMANDS.selectedLine,
      selectedLineIndex,
    });
  }

  onPressPreviousLine() {
    let { selectedLineIndex } = this.state;
    if (selectedLineIndex == null) {
      return;
    }
    selectedLineIndex = Math.max(0, selectedLineIndex - 1);
    this.setState({ selectedLineIndex });
  }

  onPressX() {
    if (
      [COMMANDS.replace, COMMANDS.replaceAll].includes(this.state.command) &&
      this.state.selectedWord != null
    ) {
      this.setState({
        command: COMMANDS.selectedWord,
      });
    } else if (
      [COMMANDS.insert].includes(this.state.command) &&
      this.state.selectedLineIndex != null
    ) {
      this.setState({
        command: COMMANDS.selectedWord,
      });
    } else {
      this.setState({
        command: null,
        selectedLocation: null,
        selectedWord: null,
        selectedLineIndex: null,
      });
    }
  }

  onSelectLine = (selectedLineIndex: number) => {
    const selectedLineTimestamp = +new Date();
    this.setState((oldState: State) => {
      const { data } = oldState;
      if (data == null) {
        return {};
      }
      if (
        oldState.command === COMMANDS.selectedLine &&
        oldState.selectedLineIndex === selectedLineIndex &&
        selectedLineTimestamp - (oldState.selectedLineTimestamp || 0) < 1000
      ) {
        return {
          command: COMMANDS.insert,
          selectedLineIndex,
        };
      }
      return {
        selectedLineTimestamp,
        selectedLineIndex,
        command: COMMANDS.selectedLine,
        selectedWord: null,
        selectedLocation: null,
      };
    });
  };

  onSelectWord = (word: string, location: RawLocation) => {
    this.setState({
      command: COMMANDS.selectedWord,
      selectedWord: word,
      selectedLocation: location,
      selectedLineIndex: null,
    });
  };

  onSelectSelectedWord() {
    this.setState({
      command: COMMANDS.replace,
      replacementWord: this.state.selectedWord,
    });
  }

  onPressSearch() {
    this.setState({ command: COMMANDS.search });
  }

  onPressEnterLine() {
    this.setState({ command: COMMANDS.goToLine });
  }

  onChangeSelectedWord(selectedWord: string) {
    this.setState({
      selectedWord,
      selectedLocation: null,
    });
  }

  onConfirmGoToLine(line: string) {
    const selectedLineIndex = Number(line);
    this.setState({
      selectedWord: null,
      selectedLocation: null,
      selectedLineIndex,
      command: COMMANDS.selectedLine,
    });
  }

  onChangeGoToLineText(goToLineText: string) {
    this.setState({
      goToLineText,
      selectedWord: null,
      selectedLocation: null,
    });
  }

  onConfirmSelectedWord(selectedWord: string) {
    let { selectedLocation } = this.state;
    const { data } = this.state;
    if (data == null || selectedLocation == null) {
      return;
    }

    selectedLocation = getNextLocation(selectedWord, selectedLocation, data);
    this.setState({
      command: selectedWord != null ? COMMANDS.selectedWord : null,
      selectedWord,
      selectedLocation,
    });
  }

  onChangeReplacementWord(replacementWord: string) {
    this.setState({
      replacementWord,
    });
  }

  onConfirmReplaceAll(replacementWord: string) {
    let { data } = this.state;
    const { selectedWord } = this.state;
    if (data == null || selectedWord == null) {
      return;
    }

    const regex = RegExp(selectedWord, 'g');
    // TODO handle this with ./text
    data = data.replace(regex, replacementWord);

    this.setState({
      data,
      command: COMMANDS.selectedWord,
      selectedWord: replacementWord,
      selectedLocation: null,
    });
  }

  onConfirmReplaceWord(replacementWord: string) {
    let { data } = this.state;
    const { selectedLocation } = this.state;
    if (data == null || selectedLocation == null) {
      return;
    }
    // TODO handle this with ./text
    data = getDataWithReplaceWord(replacementWord, selectedLocation, data);

    this.setState({
      data,
      command: COMMANDS.selectedWord,
      replacementWord,
    });
  }

  onPressReplace() {
    this.setState({
      command: COMMANDS.replace,
    });
  }

  onPressReplaceAll() {
    this.setState({
      command: COMMANDS.replaceAll,
    });
  }

  onPressEdit() {
    const { data, selectedLineIndex } = this.state;
    if (data == null || selectedLineIndex == null) {
      return;
    }
    this.setState({
      command: COMMANDS.insert,
    });
  }

  controlBarActions = {
    onPressNextSelection: () => this.onPressNextSelection(),
    onPressPreviousSelection: () => this.onPressPreviousSelection(),
    onPressX: () => this.onPressX(),
    onPressSearch: () => this.onPressSearch(),
    onChangeSelectedWord: (t: string) => this.onChangeSelectedWord(t),
    onPressEdit: () => this.onPressEdit(),
    onPressReplace: () => this.onPressReplace(),
    onPressReplaceAll: () => this.onPressReplaceAll(),
    onPressEnterLine: () => this.onPressEnterLine(),
    onPressPreviousLine: () => this.onPressPreviousLine(),
    onPressNextLine: () => this.onPressNextLine(),
    onPressInsertDone: () => this.onPressInsertDone(),
  };

  renderTextInputBar() {
    let actions = null;
    let label: ?string = null;
    let keyboardType = 'default';
    switch (this.state.command) {
      case COMMANDS.search:
        actions = {
          onChangeText: word => this.onChangeSelectedWord(word),
          onConfirmText: word => this.onConfirmSelectedWord(word),
        };
        label = this.state.selectedWord;
        break;
      case COMMANDS.goToLine:
        actions = {
          onChangeText: line => this.onChangeGoToLineText(line),
          onConfirmText: line => this.onConfirmGoToLine(line),
        };
        keyboardType = 'numeric';
        label = this.state.goToLineText;
        break;
      case COMMANDS.replace:
        actions = {
          onChangeText: word => this.onChangeReplacementWord(word),
          onConfirmText: word => this.onConfirmReplaceWord(word),
        };
        label = this.state.replacementWord;
        break;
      case COMMANDS.replaceAll:
        actions = {
          onChangeText: word => this.onChangeReplacementWord(word),
          onConfirmText: word => this.onConfirmReplaceAll(word),
        };
        label = this.state.replacementWord;
        break;
      default:
        break;
    }
    // if there are no actions to take, then we shouldn't render the bar
    if (actions === null || label == null) {
      return null;
    }
    return (
      <TextInputBar text={label} keyboardType={keyboardType} {...actions} />
    );
  }

  render() {
    return (
      <KeyboardAvoidingView style={[this.props.style]} behavior="padding">
        <Buffer
          ref={this.captureBufferRef}
          onSelectWord={this.onSelectWord}
          onSelectLine={this.onSelectLine}
          onChangeData={this.onChangeData}
          style={styles.buffer}
          width={this.props.dimensions.width}
          isEditing={this.state.command === COMMANDS.insert}
          data={this.state.data}
          selectedWord={this.state.selectedWord}
          selectedLineIndex={this.state.selectedLineIndex}
          selectedLocation={this.state.selectedLocation}
        />
        {this.renderTextInputBar()}
        <ControlBar
          {...this.props}
          {...this.state}
          actions={this.controlBarActions}
          style={styles.controlBar}
        />
      </KeyboardAvoidingView>
    );
  }
}
