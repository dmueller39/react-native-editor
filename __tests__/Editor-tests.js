// @flow
import 'react-native';
import React from 'react';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

import Editor from '../Editor';

it('renders correctly', () => {
  const tree = renderer.create(
    <Editor data="foo" dimensions={{ width: 320 }} onUpdateData={() => {}} />
  );

  expect(tree).toMatchSnapshot();
});
