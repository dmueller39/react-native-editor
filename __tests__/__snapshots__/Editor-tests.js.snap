// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  keyboardVerticalOffset={0}
  onLayout={[Function]}
  style={
    Array [
      Array [
        null,
      ],
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "continued": false,
          "continuing": false,
          "end": 3,
          "isEditing": false,
          "isSelected": false,
          "rawLineIndex": 0,
          "start": 0,
          "text": "foo",
          "textSections": Array [
            Object {
              "end": 3,
              "highlight": "normal",
              "start": 0,
              "text": "foo",
            },
          ],
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    getItemLayout={[Function]}
    horizontal={false}
    initialNumToRender={10}
    isEditing={false}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onChangeData={[Function]}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onSelectLine={[Function]}
    onSelectWord={[Function]}
    onViewableItemsChanged={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    selectedLineIndex={undefined}
    selectedLocation={undefined}
    selectedWord={undefined}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "overflow": "hidden",
      }
    }
    updateCellsBatchingPeriod={50}
    width={320}
    windowSize={5}
  >
    <View>
      <View
        onLayout={undefined}
        style={null}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={undefined}
          testID={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {},
                Object {
                  "fontFamily": "Courier New",
                  "fontSize": 14,
                },
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              disabled={false}
              ellipsizeMode="tail"
              style={Object {}}
            >
              foo
            </Text>
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "borderTopColor": "#AAAAAA",
        "borderTopWidth": 1,
        "flexDirection": "row",
        "height": 44,
        "paddingLeft": 10,
        "paddingRight": 10,
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
          },
          Object {
            "alignSelf": "center",
          },
        ]
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "fontFamily": "Courier New",
            "minWidth": 44,
            "paddingBottom": 10,
            "paddingLeft": 10,
            "paddingRight": 10,
            "paddingTop": 10,
            "textAlign": "center",
          }
        }
      >
        /abc/
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
          },
          Object {
            "alignSelf": "center",
          },
        ]
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "fontFamily": "Courier New",
            "minWidth": 44,
            "paddingBottom": 10,
            "paddingLeft": 10,
            "paddingRight": 10,
            "paddingTop": 10,
            "textAlign": "center",
          }
        }
      >
        #
      </Text>
    </View>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </View>
</View>
`;
